app/src/main/java/com/github/bsideup/liiklus/service/LiiklusService.java

```diff
@@ -83,6 +85,42 @@
                 .log("publish", Level.SEVERE, SignalType.ON_ERROR);
     }
 
+    private static Envelope toEnvelope(PublishRequest request) {
+        var eventCase = request.getEventCase();
+        switch (eventCase) {
+            case EVENT_NOT_SET:
+                return new Envelope(
+                        request.getTopic(),
+                        request.getKey().asReadOnlyByteBuffer(),
+                        request.getValue().asReadOnlyByteBuffer()
+                );
+            case LIIKLUSEVENT:
+                var cloudEvent = request.getLiiklusEvent();
+                var cloudEventBuilder = CloudEventBuilder.builder()
+                        .withId(cloudEvent.getId())
+                        .withType(cloudEvent.getType())
+                        .withSource(URI.create(cloudEvent.getSource()))
+                        .withDataContentType(cloudEvent.getDataContentType())
+                        .withData(cloudEvent.getData().toByteArray());
+
+                if (!StringUtils.isEmpty(cloudEvent.getTime())) {
+                    cloudEventBuilder = cloudEventBuilder.withTime(
+                            ZonedDateTime.parse(cloudEvent.getTime())
+                    );
+                }
+
+                return new Envelope(
+                        request.getTopic(),
+                        request.getKey().asReadOnlyByteBuffer(),
+                        it -> (ByteBuffer) it,
+                        cloudEventBuilder.build(),
+                        it -> ByteBuffer.wrap(Json.binaryEncode(it)).asReadOnlyBuffer()
```

Comment:
Again, this feels weird, for reasons slightly different from the in-memory case (assuming I understand what this code does from a very quick glance). We don't want to use JSON as a preferred way of storing data.

Reply:
First part of this answer explains this part:
https://github.com/bsideup/liiklus/pull/252#discussion_r373054513

Reply:
tl;dr: this function is only executed when some pre-processor asks for the `ByteBuffer` of this envelope (aka "serialize it for me"). It is there for the backward compatibility and will be removed once `CloudEvent` becomes the only type of `Envelope`'s value.